src/DotLiquid/Context.cs

```diff
@@ -461,33 +485,39 @@ private object FindVariable(string key)
         /// <returns></returns>
         private object Variable(string markup, bool notifyNotFound)
         {
-            List<string> parts = R.Scan(markup, VariableParserRegex);
+            using (var partsEnumerator = SyntaxCompatibilityLevel >= SyntaxCompatibility.DotLiquid22 ? Tokenizer.GetVariableEnumerator(markup) : R.Scan(markup, VariableParserRegex).GetEnumerator())
+            {
+                if (TryGetVariable(partsEnumerator, out var variable))
+                    return variable;
+                if (notifyNotFound)
+                    Errors.Add(new VariableNotFoundException(string.Format(Liquid.ResourceManager.GetString("VariableNotFoundException"), markup)));
+                return null;
+            }
+        }
 
+        private bool TryGetVariable(IEnumerator<string> partsEnumerator, out object variable)
+        {
             // first item in list, if any
-            string firstPart = parts.TryGetAtIndex(0);
-
-            Match firstPartSquareBracketedMatch = SquareBracketedRegex.Match(firstPart);
-            if (firstPartSquareBracketedMatch.Success)
-                firstPart = Resolve(firstPartSquareBracketedMatch.Groups[1].Value).ToString();
+            string firstPart = partsEnumerator.MoveNext() ? partsEnumerator.Current : null;
+            if (firstPart != null && firstPart[0] == '[')
+                firstPart = Resolve(firstPart.Substring(1, firstPart.Length - 2)).ToString();
 
             object @object;
-            if ((@object = FindVariable(firstPart)) == null)
+            if (firstPart == null || !TryFindVariable(firstPart, out @object))
             {
-                if (notifyNotFound)
-                    Errors.Add(new VariableNotFoundException(string.Format(Liquid.ResourceManager.GetString("VariableNotFoundException"), markup)));
-                return null;
+                variable = null;
+                return false;
             }
 
             // try to resolve the rest of the parts (starting from the second item in the list)
-            for (int i = 1; i < parts.Count; ++i)
+            while (partsEnumerator.MoveNext())
             {
-                var forEachPart = parts[i];
-                Match partSquareBracketedMatch = SquareBracketedRegex.Match(forEachPart);
-                bool partResolved = partSquareBracketedMatch.Success;
+                string forEachPart = partsEnumerator.Current;
+                bool partResolved = forEachPart[0] == '[';
 
                 object part = forEachPart;
                 if (partResolved)
-                    part = Resolve(partSquareBracketedMatch.Groups[1].Value);
+                    part = Resolve(forEachPart.Substring(1, forEachPart.Length - 2));

```

Comment:
Can the part be malformed as just "[" here and lead to an exception because length is 1 so the input becomes -1?

Reply:
tokenizer would throw an exception in that case. [ must be ended with ] and [" must end with "] - by itself it is not a word-character and not a complete token. Regex expression also works this way (but in that case, it would ignore the [ and not return as part of any token).